lib/galaxy/jobs/runners/condor.py

```diff
@@ -52,6 +52,24 @@ class CondorJobRunner(AsynchronousJobRunner):
 
     runner_name = "CondorRunner"
 
+    def __init__(self, app, nworkers, **kwargs):
+        """Start the job runner."""
+        runner_param_specs = {
+            "prefix": dict(map=str, default=""),
+            # String to prepend to HTCondor command line tool invocations. Defaults to an empty string.
+            "condor_rm_cmd": dict(map=str, default="condor_rm"),
```

Comment:
what are you doing here with the dict and the map?

Reply:
Runner parameters (`runner_params`) are an instance of [`RunnerParams`](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/jobs/runners/__init__.py#L76). `RunnerParams` inherits from [`ParamsWithSpecs`](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/util/__init__.py#L1279).

The point of `ParamsWithSpecs` is to enforce restrictions (`specs`) on some data (`params`). Have a a look [at the code](https://github.com/kysrpex/galaxy/blob/78426198f1816bdb7328e4d73b40855a9c76bc09/lib/galaxy/util/__init__.py#L1285-L1295) for greater details. In few words, it disallows keys not defined in the spec, it can process/coerce the values using the `map` function, and reject invalid values using a function `valid` that determines their validity. Specs are **not optional**, and seem to be expected to be a [`Mapping`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping).